### PR TITLE
Make Mapping::clone return a unique_ptr<Mapping>.

### PR DESCRIPTION
**Pull request by drwells**
This commit converts the current interface into one based on smart pointers. It also gets rid of the use of a covariant return type in MappingQ1::clone and MappingQ1Eulerian::clone.

Part of #5880.

### CODE REVIEW

**Comment by drwells** [comment]
It looks like I forgot quite a few of those. This is now fixed.

(one problem with the unity build is it makes it very easy to forget these headers)

File: source/fe/mapping_q.cc [hunk 1]
```diff
@@ -88,18 +89,33 @@ template <int dim, int spacedim>
...
+{
+  // Note that we really do have to use clone() here, since mapping.q1_mapping
+  // may be MappingQ1Eulerian and mapping.qp_mapping may be MappingQEulerian.
+  std::unique_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
+  const auto q1_map = dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_q1_map.get());
+  Assert(q1_map != nullptr, ExcInternalError());
+  Assert(q1_map->get_degree() == 1, ExcInternalError());
+  q1_mapping.reset(q1_map);
+  // avoid double delete
+  other_q1_map.release();
```
Comment (masterleinad) [inline source/fe/mapping_q.cc]: Calling `release()` without assigning always looks really dangerous to me.
I don't see a reason why we can't do
```
std::unique_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
q1_mapping.reset(dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_q1_map.release()));
Assert(q1_mapping, ExcInternalError());
Assert(q1_mapping->get_degree() == 1, ExcInternalError());
```
directly avoiding the unassigned `release()` call and the local variable `q1_map`.

File: source/fe/mapping_q.cc [hunk 2]
```diff
@@ -88,18 +89,33 @@ template <int dim, int spacedim>
...
+      qp_mapping = q1_mapping;
+    }
+  else
+    {
+      std::unique_ptr<Mapping<dim,spacedim>> other_qp_map = mapping.qp_mapping->clone();
+      const auto qp_map = dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_qp_map.get());
+      Assert(qp_map != nullptr, ExcInternalError());
+      qp_mapping.reset(qp_map);
+      // avoid double delete
+      other_qp_map.release();
```
Comment (masterleinad) [inline source/fe/mapping_q.cc]: Similarly as above, I would write this as:
```
std::unique_ptr<Mapping<dim,spacedim>> other_qp_map = mapping.qp_mapping->clone();
qp_mapping.reset(dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_qp_map.release()));
Assert(qp_mapping, ExcInternalError());
```

File: source/fe/mapping_q.cc (hunk 1 above)
Comment (drwells) [reply source/fe/mapping_q.cc]: Thats nicer than what I did. I'll update it.

**Comment by drwells** [comment]
I am a little surprised that you brought up `Winconsistent-missing-override`. I get an incredible number of warnings when I compile with this flag on clang (which is why I have long since disabled it). I don't mind fixing these, but a lot are unrelated to this patch (which I also don't mind fixing; I need something to help me procrastinate on grading exams today).

Do you get lots of other warnings with this flag enabled that are unrelated to this patch?

**Comment by masterleinad** [comment]
> Do you get lots of other warnings with this flag enabled that are unrelated to this patch?

No, I don't. Actually, I didn't enable this flag explicitly.

**Comment by drwells** [comment]
I believe that I fixed the inconsistent override warning. I also wrote a bigger patch that adds `override` everywhere that clang-tidy suggests; I will finish that PR at a later time.

I am no longer confident that the proposed fix cannot leak. If we do:
```cpp
std::unique_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
q1_mapping.reset(dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_q1_map.release()));
```
then the following can occur:
1. at the moment `std::unique_ptr::release` finishes, no object has been constructed that will call `delete` on the pointer to the cloned mapping.
2. If the `dynamic_cast` fails then no object stores a pointer to the cloned mapping.
3. We throw an exception in the next line. We get a memory leak and an internal error (which will almost surely stop the program, so maybe leaks don't matter).

The approach I wrote is clunkier but I believe that we must call `release` last to avoid possible leaks. Is my interpretation correct?

**Comment by masterleinad** [comment]
> I believe that I fixed the inconsistent override warning. I also wrote a bigger patch that adds override everywhere that clang-tidy suggests; I will finish that PR at a later time.

Great! I had some trouble with reintroducing the `virtual` when I last tried.

**Comment by masterleinad** [comment]
> The approach I wrote is clunkier but I believe that we must call release last to avoid possible leaks. Is my interpretation correct?

Yes, I agree. What about calling `get()` in the assignment (`reset`) and `release()` after the first `Assert`? Since `release` is `noexcept` that should be fine and is not much more complicated (we still avoid the other local variable).

**Comment by drwells** [comment]
> Great! I had some trouble with reintroducing the `virtual` when I tried.

I also had a lot of trouble: I added `virtual` back in with a lot of patience and several emacs macros. I am not convinced that I managed to mark everything as `override` that could be marked as `override` so I will keep working on it.

> What about calling `get()` in the assignment (`reset`) and `release()` after the first `Assert`?

Hm. That can still result in a memory leak unless we guard the `nullptr` check, i.e., we should do
```cpp
if (q1_mapping == nullptr)
  Assert(q1_mapping != nullptr, ExcInternalError());
else
  other_q1_map.release();
```
again in the (internal failure) case that `dynamic_cast` fails. Is that agreeable? Its not that much better than my original version :(

**Comment by masterleinad** [comment]
The code after we fail does not really matter, it is just important that we didn't release ownership of the pointer before. Therefore, I think that we are good even without the additional guard (unless we are in `Release` mode).
If you want to handle this correctly also in `Release` mode, I would prefer the original version indeed.

**Comment by bangerth** [comment]
Here is an example of how I think the `dynamic_cast` of `std::unique_ptr` should be handled: https://stackoverflow.com/questions/21174593/downcasting-unique-ptrbase-to-unique-ptrderived

I don't think you can avoid an `if` statement. (I don't think you need the separate function if you use it in only one place, but the idiom they use seems sensible.)

**Comment by drwells** [comment]
@bangerth A link from that page yielded, I believe, the correct answer. We can write something nicer since we are ultimately using a `shared_ptr`:

```cpp
  std::shared_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
  q1_mapping = std::dynamic_pointer_cast<const MappingQGeneric<dim,spacedim>>
    (other_q1_map);
  Assert(q1_mapping != nullptr, ExcInternalError());
```

this is slightly less efficient since `other_q1_map`'s existence increments and then decrements the use count, but its a lot cleaner.

I think this is our best answer since it is memory safe and concise.

**Comment by masterleinad** [comment]
Yes, that looks good.

**Comment by bangerth** [comment]
/run-tests

OK to merge once the tester is happy.

**Pull request merged**